### ISSUE TITLE
does ghub work with SSO GHE?

### ISSUE THREAD

**Issue by donahchoo** [magit/ghub#54]
I'm wondering if the root of my issue is that ghub doesn't work with SSO GHE?  I've gotten hub to work, but maybe there's something in ghub that needs a tweak.  I can't get ghub-create-token to complete successfully because of our SSO redirect.  ghub ends up with:

`HTTP Error: 400, "Bad Request", "and Github didn't return JSON"`

**Comment by vermiculus** [comment]
SSO…single sign-on?

**Comment by tarsius** [comment]
Ghub does not support [Authenticating for SAML SSO](https://developer.github.com/v3/auth/#authenticating-for-saml-sso). Github's documentation on the matter are quite light. The section I have linked to is the main entry point to it. It is a subsection of "Basic Authentication", so I am assuming it is a form of just that.

Ghub does support Basic Authentication for the purpose of creating a token. Theoretically Basic Authentication can also be used for other resources but that's not part of the *Ghub* API (e.g. changes here do not result in the major version being bumped).

This probably means that you have to create the token manually. You for sure have to [authorize it manually](https://help.github.com/articles/authorizing-a-personal-access-token-for-use-with-a-saml-single-sign-on-organization/) and login to your identity provider. Ghub does not understand the errors that result from you not being logged into that, so when an error occurs you won't get a meaningful error but logging into your identity provider might make the request succeed on the next try.

I did not see anything about clients having to provide some additional request header, but if they do, then this won't work at all.

I don't have access to an enterprise edition.

TL;DR No. (Probably)

**Comment by donahchoo** [comment]
actually, I got this to work!

**Comment by vermiculus** [comment]
@donahchoo Can you share your solution here?

![image](https://imgs.xkcd.com/comics/wisdom_of_the_ancients.png)

**Comment by am** [comment]
@vermiculus that made me laugh 🤣

@donahchoo please, the world needs you! I'm having the same problem, our GHE auth is done using Okta for login, this means that it also have two-factor authentication enabled. Still didn't figure out how can this be achieved...

**Comment by donahchoo** [comment]
@vermiculus @am I think I did this: https://mrdias.com/2018/04/01/using-magit-with-github-enterprise.html

however, my ghub is currently busted so I can't be sure.

**Comment by am** [comment]
@donahchoo thank you for getting back and share your solution. From what I can see in the article the solution uses `.authinfo` to set a password, AFAIK this won't work with SSO.

**Comment by donahchoo** [comment]
@am it works for my SSO.  Our GHE is behind SSO.  We auth to SSO, access GHE and generate a token from GHE that goes into .authinfo.  seems to work here, now that I got ghub and forge setup correctly.
